src/scheduler/scheduler.rs

```diff
@@ -80,36 +85,61 @@ impl<VM: VMBinding> GCWorkScheduler<VM> {
         self.worker_group.as_ref().worker_count()
     }
 
-    /// Create GC threads, including the controller thread and all workers.
+    /// Create GC threads for the first time.  It will also create the `GCWorker` instances.
+    ///
+    /// Currently GC threads only include worker threads, and we currently have only one worker
+    /// group.  We may add more worker groups in the future.
     pub fn spawn_gc_threads(self: &Arc<Self>, mmtk: &'static MMTK<VM>, tls: VMThread) {
-        // Spawn the controller thread.
-        let coordinator_worker = GCWorker::new(
-            mmtk,
-            usize::MAX,
-            self.clone(),
-            true,
-            self.coordinator_worker_shared.clone(),
-            deque::Worker::new_fifo(),
-        );
-        let gc_controller = GCController::new(
-            mmtk,
-            mmtk.gc_requester.clone(),
-            self.clone(),
-            coordinator_worker,
-        );
-        VM::VMCollection::spawn_gc_thread(tls, GCThreadContext::<VM>::Controller(gc_controller));
+        self.worker_group.create_workers(mmtk);
+        self.worker_group.spawn(tls)
```

Comment:
`create_workers()` sets the worker state to `Resting`, and `spawn()` immediately updates the state to `Spawned`. Looks like the `Resting` state is not necessary?  Also probably the two-step initialization can be merged.

If `create_workers` also directly spawns the workers, is it possible to get rid of the `Resting` state during this stage? If this is true, we can also rename the Resting state to `Surrendered` to only refer to the state where GC workers are waiting for forking.

Reply:
Good idea.  I started with creating `GCWorker` instances and spawning them in one step, but I found it was easier to implement by keeping them separate.  But from the perspective of state transition, it is better to make one state transition instead of two in `initialize_collection`.

`Surrendered` is also better than `Resting`, I think.  This will emphasise that `Surrendered` is a special state for forking, rather than a routine state that MMTk needs to go through for all VM bindings (It's only necessary for VMs that need forking).

I'll refactor it to change the two state transitions into one.

Reply:
Done.